### PR TITLE
can: dts: Convert can-primary alias to zephyr,can-primary chosen property

### PR DESCRIPTION
**Pull request by galak**
Convert CAN to use a chosen node property that is similar to how we
handle zephyr,entroy or zephyr,flash-controller as the means to select a
specific peripheral instance utilized by a subsystem.

Replace references of the form:

alias {
	can-primary = &can1;
};

with:

chosen {
	zephyr,can-primary = &can1;
};

Replace various macro/define references with either
DT_CHOSEN(zephyr_can) or replace DT_ALIAS_CAN_PRIMARY_LABEL with
DT_CHOSEN_ZEPHYR_CAN_PRIMARY_LABEL.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

**Review by mbolivar-nordic — COMMENTED** [review]
I have never used CAN, so a beginner question: I see the alias is named "can-primary", and the chosen node is named "zephyr,can". Is there a common enough use case for multiple CANs that would make the chosen property better named "zephyr,can-primary" or so?

**Comment by alexanderwachter** [comment]
We thought about using a chosen node when we introduced the can-primary label. We decided to not use it sine we sometimes have more than one can an maybe in the future we will have a subsys that will use can-secondary. An example would be redundancy. can-primary makes more sense imo.

**Comment by galak** [comment]
> We thought about using a chosen node when we introduced the can-primary label. We decided to not use it sine we sometimes have more than one can an maybe in the future we will have a subsys that will use can-secondary. An example would be redundancy. can-primary makes more sense imo.

We can adjust the name to be `can-0` and `can-1` or whatever, but the chosen node is the method we've decided on to convey node selection of this manor.

**Comment by alexanderwachter** [comment]
can-0 and can-1 didn‘t work because they refer to the controller. Some SoC start counting with CAN0, some start with CAN1.
That was why we introduced can-primary in the first place. Chosen can only be one node.

**Comment by galak** [comment]
> can-0 and can-1 didn‘t work because they refer to the controller. Some SoC start counting with CAN0, some start with CAN1.

So I understand the point here.

> That was why we introduced can-primary in the first place. Chosen can only be one node.

Not sure I follow what you mean by it can only be one node?  Alias can only represent one path.  So how is that different?

**Comment by henrikbrixandersen** [comment]
I am okay with moving to a chosen zephyr,xxx style for this. I do believe, however, that we should come up with a naming scheme for chosen nodes where we can have "more than one" of a given type.

For CAN, these would indicate primary/seconday (for redundancy) interfaces. I reckon something similar could be a use case for other network interfaces (e.g. ethernet)?

**Comment by alexanderwachter** [comment]
@galak what about zephyr,can-primary = &can1; ?

File: boards/arm/black_f407_generic/black_f407ve.dts [hunk 1]
```diff
@@ -17,6 +17,7 @@
 		zephyr,sram = &sram0;
 		zephyr,flash = &flash0;
 		zephyr,ccm = &ccm0;
+		zephyr,can = &can2;
```
Comment (alexanderwachter) [inline boards/arm/black_f407_generic/black_f407ve.dts:20]: Please use `zephyr,can-primary`.
Otherwise it lookf good. thanks for the PR!

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

**Comment by galak** [comment]
Updated to be can-primary.

**Review by alexanderwachter — APPROVED** [review]

File: include/devicetree/zephyr.h [hunk 2]
```diff
@@ -55,6 +64,9 @@
 	DT_LABEL(DT_CHOSEN(zephyr_flash_controller))
 #endif
 
+#if DT_NODE_HAS_PROP(DT_CHOSEN(zephyr_can), label)
```
Comment (mbolivar-nordic) [inline include/devicetree/zephyr.h:67]: missed this one

File: samples/subsys/canbus/canopen/src/main.c [hunk 3]
```diff
@@ -15,9 +15,9 @@
 LOG_MODULE_REGISTER(app);
 
 
-#define CAN_INTERFACE DT_ALIAS_CAN_PRIMARY_LABEL
-#define CAN_BITRATE (DT_ALIAS_CAN_PRIMARY_BUS_SPEED / 1000)
-#if !defined(DT_ALIAS_CAN_PRIMARY_LABEL)
+#define CAN_INTERFACE DT_CHOSEN_ZEPHYR_CAN_PRIMARY_LABEL
+#define CAN_BITRATE (DT_PROP(DT_CHOSEN(zephyr_can), bus_speed) / 1000)
```
Comment (mbolivar-nordic) [inline samples/subsys/canbus/canopen/src/main.c:19]: and here

File: include/devicetree/zephyr.h (hunk 2 above)
Comment (galak) [reply include/devicetree/zephyr.h]: fixed

File: samples/subsys/canbus/canopen/src/main.c (hunk 3 above)
Comment (galak) [reply samples/subsys/canbus/canopen/src/main.c]: fixed

**Review by mbolivar-nordic — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Pull request merged**